File: typescript/packages/core/src/core/gsheets/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GSheetsAccessor } from '../../accessor/gsheets.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { enoent } from '../../utils/errors.ts'
import { sheetsBase, type TokenManager, googleGet } from '../google/client.ts'
import { resolveEntry } from '../hierarchy/probe.ts'
import { makeRead } from '../hierarchy/read.ts'
import type { ScopeMatch } from '../hierarchy/scope.ts'
import { compactJsonBytes } from '../render/json.ts'
import { readdir } from './readdir.ts'
import { detectScope } from './scope.ts'

const GRID_DATA_PARAM = 'true'

/**
 * Fetch full spreadsheet JSON, cell values included.
 *
 * `spreadsheets.get` returns no grid data unless asked, so without
 * `includeGridData` the rendered `.gsheet.json` is tab metadata and nothing
 * an agent can read a cell from.
 */
export async function readSpreadsheet(
  tm: TokenManager,
  spreadsheetId: string,
): Promise<Uint8Array> {
  const url = `${sheetsBase(tm)}/spreadsheets/${spreadsheetId}`
  const data = await googleGet(tm, url, { includeGridData: GRID_DATA_PARAM })
  return compactJsonBytes(data)
}

export async function readValues(
  tm: TokenManager,
  spreadsheetId: string,
  range: string,
): Promise<Uint8Array> {
  const url = `${sheetsBase(tm)}/spreadsheets/${spreadsheetId}/values/${range}`
  const data = await googleGet(tm, url)
  return compactJsonBytes(data)
}

async function readFile(
  accessor: GSheetsAccessor,
  _match: ScopeMatch,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const entry = await resolveEntry(readdir, accessor, path, index)
  if (entry === null) throw enoent(path.virtual)
  return readSpreadsheet(accessor.tokenManager, entry.id)
}

export const read = makeRead<GSheetsAccessor>(detectScope, { file: readFile })

export async function* stream(
  accessor: GSheetsAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  yield await read(accessor, path, index)
}

File: python/mirage/core/gsheets/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.gsheets import GSheetsAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gsheets.client import TokenManager, google_get, sheets_base
from mirage.core.gsheets.readdir import readdir
from mirage.core.gsheets.scope import detect_scope
from mirage.core.hierarchy.probe import resolve_entry
from mirage.core.hierarchy.read import make_read
from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.render.json import compact_json_bytes
from mirage.types import PathSpec
from mirage.utils.errors import enoent

GRID_DATA_PARAM = "true"


async def read_spreadsheet(token_manager: TokenManager,
                           spreadsheet_id: str) -> bytes:
    """Fetch full spreadsheet JSON, cell values included.

    `spreadsheets.get` returns no grid data unless asked, so without
    `includeGridData` the rendered `.gsheet.json` is tab metadata and
    nothing an agent can read a cell from.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.

    Returns:
        bytes: JSON response as bytes.
    """
    url = f"{sheets_base(token_manager)}/spreadsheets/{spreadsheet_id}"
    data = await google_get(token_manager, url,
                            {"includeGridData": GRID_DATA_PARAM})
    return compact_json_bytes(data)


async def read_values(token_manager: TokenManager, spreadsheet_id: str,
                      range_: str) -> bytes:
    """Read cell values via Values API. Returns JSON array.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.
        range_ (str): A1 notation range.

    Returns:
        bytes: JSON response as bytes.
    """
    base = sheets_base(token_manager)
    url = f"{base}/spreadsheets/{spreadsheet_id}/values/{range_}"
    data = await google_get(token_manager, url)
    return compact_json_bytes(data)


async def _read_file(accessor: GSheetsAccessor, match: ScopeMatch,
                     path: PathSpec, index: IndexCacheStore) -> bytes:
    entry = await resolve_entry(readdir, accessor, path, index)
    if entry is None:
        raise enoent(path.virtual)
    return await read_spreadsheet(accessor.token_manager, entry.id)


read = make_read(detect_scope, readers={"file": _read_file})
